src/cryptonote_core/blockchain.cpp

```diff
@@ -3552,6 +3553,10 @@ bool Blockchain::handle_block_to_main_chain(const block& bl, const crypto::hash&
   get_difficulty_for_next_block(); // just to cache it
   invalidate_block_template_cache();
 
+  std::shared_ptr<tools::Notify> block_notify = m_block_notify;
```

Comment:
@moneromooo-monero @jtgrassie 
Regarding https://github.com/monero-project/monero/pull/4333#discussion_r221421359

Had to revisit the standard once again. Unfortunately, yes, `std::shared_ptr` copy constructor is not atomic. There are `std::atomic_load` and `std::atomic_store` overloads for this case in c++11. 

But they are going to be deprecated in c++20 and the semantic will be changed to `std::atomic<std::shared_ptr<T>>`. http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0718r2.html#3.3

I bet we don't want to deal with different compilation issues tied to these changes.

We have no choice but to introduce kind of manual locking.

Reply:
This is set once at program start though. In practice it can't race.